python/pyspark/sql/functions.py

```diff
@@ -3514,7 +3538,19 @@ def map_from_arrays(col1: "ColumnOrName", col2: "ColumnOrName") -> Column:
     return Column(sc._jvm.functions.map_from_arrays(_to_java_column(col1), _to_java_column(col2)))
 
 
+@overload
 def array(*cols: "ColumnOrName") -> Column:
+    ...
+
+
+@overload
+def array(__cols: Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]) -> Column:
+    ...
+
+
+def array(
+    *cols: Union["ColumnOrName", Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]]
```

Comment:
Out of curiosity would flattening unions of unions explicitly be better?

Reply:
In general, nested `Unions` are flattened automatically:

```python
>>>  Union["ColumnOrName", Union[List["ColumnOrName"], Tuple["ColumnOrName", ...]]]
typing.Union[ForwardRef('ColumnOrName'), typing.List[ForwardRef('ColumnOrName')], typing.Tuple[ForwardRef('ColumnOrName'), ...
```

Personally, I find structure that clearly groups similar categories useful (for similar problem see https://github.com/apache/spark/pull/34671#discussion_r754532969), but I am not very attached to this idea.

Reply:
Makes sense. Thanks!